jgrapht-core/src/main/java/org/jgrapht/alg/tour/NearestInsertionHeuristicTSP.java

```diff
@@ -0,0 +1,324 @@
+        return new Closest<>(tourVertex, closest, minDist);
+    }
+
+    /**
+     * Update the Map storing closest unvisited vertex for each tour vertex
+     *
+     * @param currentClosest Map storing closest unvisited vertex for each tour
+     * vertex
+     * @param chosen Latest vertex added to tour
+     * @param unvisited Set of unvisited vertices
+     * @param graph The graph
+     */
+    private void updateClosest(Map<V, Closest<V>> currentClosest, Closest<V> chosen, Set<V> unvisited, Graph<V, E> graph) {
+        // Update the set of unvisited vertices, and exit if none remain
+        unvisited.remove(chosen.getUnvisitedVertex());
+        if (unvisited.isEmpty()) {
+            currentClosest.clear();
+            return;
+        }
+        // Update any entries impacted by the choice of new vertex
+        currentClosest.replaceAll((v, c) -> {
+            if (chosen.getTourVertex().equals(v) || chosen.getUnvisitedVertex().equals(c.getUnvisitedVertex())) {
+                return getClosest(v, unvisited, graph);
+            }
+            return c;
+        });
+        currentClosest.put(chosen.getUnvisitedVertex(), getClosest(chosen.getUnvisitedVertex(), unvisited, graph));
+    }
+
+    /**
+     * Chooses the closest unvisited vertex to the sub-tour
+     *
+     * @param closestVertices Map storing closest unvisited vertex for each tour
+     * vertex
+     * @return First result of sorting values
+     */
+    private Closest<V> chooseClosest(Map<V, Closest<V>> closestVertices) {
+        return Collections.min(closestVertices.values());
+    }
+
+    /**
+     * Augment an existing tour to give a complete tour
+     *
+     * @param subtour The vertices of the existing tour
+     * @param graph The graph
+     * @return List of vertices representing the complete tour
+     */
+    private List<V> augment(List<V> subtour, Graph<V, E> graph) {
+        Set<V> unvisited = new HashSet<>();
+        unvisited.addAll(graph.vertexSet());
+        unvisited.removeAll(subtour);
+        return augment(subtour, getClosest(subtour, unvisited, graph), unvisited, graph);
+    }
+
+    /**
+     * Augment an existing tour to give a complete tour
+     *
+     * @param subtour The vertices of the existing tour
+     * @param closestVertices Map of data for closest unvisited vertices
+     * @param unvisited Set of unvisited vertices
+     * @param graph The graph
+     * @return List of vertices representing the complete tour
+     */
+    private List<V> augment(List<V> subtour, Map<V, Closest<V>> closestVertices, Set<V> unvisited, Graph<V, E> graph) {
+        while (!unvisited.isEmpty()) {
+            // Select a city not in the subtour, having the shortest distance to any one of the cities in the subtoor.
+            Closest<V> closestVertex = chooseClosest(closestVertices);
+
+            // Determine the vertices either side of the selected tour vertex
+            int i = subtour.indexOf(closestVertex.getTourVertex());
```

Comment:
It might make sense to store the tour as a map `Map<V, Pair<E, E>>` from vertices to the 2 neighbouring edges so that we can retrieve previous and next vertices in the tour in expected constant time.